Add gruvbox theme

Allyedge opened the pull request:
Added the `gruvbox` which I saw was not there yet (it is one of my favourite themes 😄).

Am unsure if something is missing or should be refactored, so feedback is appreciated.

IroncladDev commented:
Thanks a ton, was planning to add Gruvbox but it looks like most of the work is already done.

I'm in the progress of refactoring some theme stuff, so I'm going to put this PR on hold and let you know once I've made the design decision for how a single package can provide multiple themes.

Allyedge commented:
Very happy to hear that, no problem.

Will be waiting! 🫡

IroncladDev commented:
Alrighty, just merged https://github.com/webtui/webtui/pull/34 which fixes #31

You should be good to go now

You will need to resolve the conflict and update the theme to be scoped with the `[data-webtui-theme="*"]` attribute

Go ahead and revert the changes for `ThemeSwitcher`, I'm going to change the way themes are displayed across the site soon

Allyedge commented:
Reverted the changes to `ThemeSwitcher` and updated the attributes.

A double check on the colors as well as their naming would be good, just to make sure they match the planned theme.

IroncladDev commented on packages/theme-gruvbox/src/base.css:
```diff
@@ -0,0 +1,92 @@
+@layer utils {
+  :root,
```
remove styling from `:root` so it's easier to switch themes dynamically

IroncladDev commented on packages/theme-gruvbox/src/base.css:
```diff
@@ -0,0 +1,92 @@
+@layer utils {
+  :root,
+  [data-webtui-theme="dark"] {
```
This overrides the default dark theme that webtui comes with. Use `[data-webtui-theme="gruvbox-dark"]` or something similar

IroncladDev commented on packages/theme-gruvbox/src/base.css:
```diff
@@ -0,0 +1,92 @@
+    --foreground0: var(--gb-dark-fg1);
+    --foreground1: var(--gb-dark-fg2);
+    --foreground2: var(--gb-dark-fg3);
+    --foreground-bright: var(--gb-dark-fg0);
```
`foreground-bright` is not part of the base theme palette. Only three foreground levels are needed, `foreground0` being the strongest in contrast and `foreground2` being the weakest

IroncladDev commented on packages/theme-gruvbox/src/base.css:
```diff
@@ -0,0 +1,92 @@
+    --accent-gray: var(--gb-dark-gray);
+  }
+
+  .gruvbox-light,
```
Remove the `.gruvbox-light` class from this selector

IroncladDev commented on packages/theme-gruvbox/src/base.css:
```diff
@@ -0,0 +1,92 @@
+  }
+
+  .gruvbox-light,
+  [data-webtui-theme="light"] {
```
```suggestion
  [data-webtui-theme="gruvbox-light"] {
```

IroncladDev commented on packages/theme-gruvbox/src/base.css:
```diff
@@ -0,0 +1,92 @@
+    --foreground2: var(--gb-dark-fg3);
+    --foreground-bright: var(--gb-dark-fg0);
+
+    --accent-red: var(--gb-red);
+    --accent-green: var(--gb-green);
+    --accent-yellow: var(--gb-yellow);
+    --accent-blue: var(--gb-blue);
+    --accent-purple: var(--gb-purple);
+    --accent-aqua: var(--gb-aqua);
+    --accent-orange: var(--gb-orange);
+    --accent-gray: var(--gb-dark-gray);
```
Seems unnecessary since there are already gruvbox-specific color vars.

`--accent-*` is not part of the base webtui theme

IroncladDev commented on packages/theme-gruvbox/src/components/badge.css:
```diff
@@ -0,0 +1,118 @@
+@layer components {
+  [is-~="badge"] {
+    &[variant-="bg0"] {
+      --badge-color: var(--background0);
+      --badge-text: var(--foreground0);
+    }
+
+    &[variant-="bg1"] {
+      --badge-color: var(--background1);
+      --badge-text: var(--foreground0);
+    }
+
+    &[variant-="bg2"] {
+      --badge-color: var(--background2);
```
Badges already have background and foreground variants (see https://webtui.ironclad.sh/components/badge/#variant)

All you need to do is add the accents

IroncladDev commented on packages/theme-gruvbox/src/components/badge.css:
```diff
@@ -0,0 +1,118 @@
+      --badge-text: var(--background0);
+    }
+
+    &[variant-="gray"] {
+      --badge-color: var(--accent-gray);
+      --badge-text: var(--background0);
+    }
+
+    &[variant-="red"] {
+      --badge-color: var(--accent-red);
+      --badge-text: var(--background0);
+    }
+
+    &[variant-="green"] {
+      --badge-color: var(--accent-green);
```
Use the `--gb-*` variables you defined instead of defining accent colors again. It's easier for users to use the names of the original theme tokens without any renaming.

IroncladDev commented on packages/theme-gruvbox/src/components/button.css:
```diff
@@ -0,0 +1,118 @@
+@layer components {
+  button {
+    &[variant-="bg0"] {
+      --button-primary: var(--background0);
+      --button-secondary: var(--foreground0);
+    }
+
+    &[variant-="bg1"] {
+      --button-primary: var(--background1);
```
background/foreground variants unnecessary

IroncladDev commented on packages/theme-gruvbox/src/components/typography.css:
```diff
@@ -0,0 +1,28 @@
+  h4,
+  h5,
+  h6 {
+    color: var(--accent-green);
```
See above comment on using original token names

IroncladDev reviewed (changes requested):
Added some comments. 

Please also add a documentation page under `web/src/pages/plugins/` and add gruvbox to `web/src/pages/start/plugins.md`.

It would be nice if you could support the hard/medium/soft gruvbox theme variants. See `packages/theme-catppuccin/base.css` on how I'm using [CSS attribute selectors](https://developer.mozilla.org/en-US/docs/Web/CSS/Attribute_selectors) to determine which catppuccin flavor is being selected

Allyedge commented:
Silly mistake with the linter there. Everything should be good now.

Updated the code to match the requests in the comments.

IroncladDev reviewed (changes requested):
Dark mode looks and works great, light mode isn't really working at all

https://github.com/user-attachments/assets/e481aba2-feb2-4173-ba55-b556718fb8b7

Other than that, just requested one common change across three files.

After you get those addressed, the gruvbox theme should be good to go.

Allyedge commented:
Updated the syntax and fixed the light mode issues.

IroncladDev commented:
LGTM
